<a href="https://colab.research.google.com/github/mmkeyes140/mmkeyes140/blob/main/spam_detector.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Spam Detector

In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

## Retrieve the Data

The data is located at [https://static.bc-edx.com/mbc/ai/m4/datasets/spam-data.csv](https://static.bc-edx.com/mbc/ai/m4/datasets/spam-data.csv)

Dataset Source: [UCI Machine Learning Library](https://archive-beta.ics.uci.edu/dataset/94/spambase)

Import the data using Pandas. Display the resulting DataFrame to confirm the import was successful.

In [ ]:
# Import the data
data = pd.read_csv("https://static.bc-edx.com/mbc/ai/m4/datasets/spam-data.csv")
data.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  spam  
0                       278     1  
1                      1028     1  
2                      2259     1  
3                       191     1  
4                       191     1  

[5 rows x 58 columns]

## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct!

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.

ANSWER - Random Forests should be more accurate.  Logistic Regression performs 1 check, Random Forests performs multiple and removes fails.

## Split the Data into Training and Testing Sets

In [ ]:
from pandas.core.frame import DataFrame
# Create the labels set `y` and features DataFrame `X`
X = data.copy()
X = X.drop(columns="spam")
X.head()
y = pd.Series(data['spam'])
y.head()

0    1
1    1
2    1
3    1
4    1
Name: spam, dtype: int64

In [ ]:
# Check the balance of the labels variable (`y`) by using the `value_counts` function.
y.value_counts()


0    2788
1    1813
Name: spam, dtype: int64

In [ ]:
# Split the data into X_train, X_test, y_train, y_test
X_train, X_test, y_train, y_test = train_test_split(X,y)
X_train.head()


word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
4276            0.00               1.42           0.00           0.0   
256             0.00               0.00           1.42           0.0   
3629            0.00               0.00           0.00           0.0   
3309            0.00               0.00           0.57           0.0   
772             0.49               0.00           0.74           0.0   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
4276           0.00            0.00              0.00                0.00   
256            0.71            0.00              0.00                0.71   
3629           0.00            0.00              0.00                0.00   
3309           0.85            0.00              0.00                0.00   
772            0.24            0.24              0.24                0.49   

      word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
4276             0.00            1.42  ...                   0.0         0.00   
256              0.00            0.71  ...                   0.0         0.00   
3629             0.00            0.00  ...                   0.0         0.00   
3309             0.28            0.00  ...                   0.0         0.08   
772              0.00            0.00  ...                   0.0         0.00   

      char_freq_(  char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
4276        0.000          0.0        0.000        0.000          0.0   
256         0.000          0.0        0.931        0.000          0.0   
3629        0.000          0.0        0.000        0.000          0.0   
3309        0.000          0.0        0.282        0.242          0.0   
772         0.035          0.0        0.315        0.035          0.0   

      capital_run_length_average  capital_run_length_longest  \
4276                       3.555                          21   
256                       12.055                         117   
3629                       1.000                           1   
3309                       2.460                          26   
772                        4.071                          60   

      capital_run_length_total  
4276                        96  
256                        217  
3629                         5  
3309                       278  
772                        228  

[5 rows x 57 columns]

## Scale the Features

Use the `StandardScaler` to scale the features data. Remember that only `X_train` and `X_test` DataFrames should be scaled.

In [ ]:
from sklearn.preprocessing import StandardScaler

# Create the StandardScaler instance
scaler = StandardScaler()


In [ ]:
# Fit the Standard Scaler with the training data
X_train_scaled = X_train.copy()
columns_to_scale = ["capital_run_length_average", "capital_run_length_longest", "capital_run_length_total"]
X_train_scaled = X_train_scaled.drop(columns = ['capital_run_length_average', 'capital_run_length_longest', 'capital_run_length_total'])
X_train_scaled[columns_to_scale] = scaler.fit_transform(X_train[columns_to_scale])
X_train_scaled.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
4276            0.00               1.42           0.00           0.0   
256             0.00               0.00           1.42           0.0   
3629            0.00               0.00           0.00           0.0   
3309            0.00               0.00           0.57           0.0   
772             0.49               0.00           0.74           0.0   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
4276           0.00            0.00              0.00                0.00   
256            0.71            0.00              0.00                0.71   
3629           0.00            0.00              0.00                0.00   
3309           0.85            0.00              0.00                0.00   
772            0.24            0.24              0.24                0.49   

      word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
4276             0.00            1.42  ...                   0.0         0.00   
256              0.00            0.71  ...                   0.0         0.00   
3629             0.00            0.00  ...                   0.0         0.00   
3309             0.28            0.00  ...                   0.0         0.08   
772              0.00            0.00  ...                   0.0         0.00   

      char_freq_(  char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
4276        0.000          0.0        0.000        0.000          0.0   
256         0.000          0.0        0.931        0.000          0.0   
3629        0.000          0.0        0.000        0.000          0.0   
3309        0.000          0.0        0.282        0.242          0.0   
772         0.035          0.0        0.315        0.035          0.0   

      capital_run_length_average  capital_run_length_longest  \
4276                   -0.049605                   -0.227404   
256                     0.208709                    0.524211   
3629                   -0.127251                   -0.383991   
3309                   -0.082882                   -0.188258   
772                    -0.033924                    0.077939   

      capital_run_length_total  
4276                 -0.317853  
256                  -0.113067  
3629                 -0.471865  
3309                 -0.009828  
772                  -0.094450  

[5 rows x 57 columns]

In [ ]:
from pickle import TRUE
# Scale the training data
X_test_scaled = X_test.copy()
columns_to_scale = ["capital_run_length_average", "capital_run_length_longest", "capital_run_length_total"]
X_test_scaled = X_test_scaled.drop(columns = ['capital_run_length_average', 'capital_run_length_longest', 'capital_run_length_total'])
X_test_scaled[columns_to_scale] = scaler.fit_transform(X_test[columns_to_scale])
X_test_scaled.head()


word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
3550             0.0                0.0           0.21           0.0   
1799             0.0                0.0           0.00           0.0   
3041             0.0                0.8           0.00           0.0   
3243             0.0                0.0           0.00           0.0   
2971             0.0                0.0           0.71           0.0   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
3550            0.0             0.0               0.0                0.21   
1799            0.0             0.0               0.0                3.19   
3041            0.6             0.0               0.0                0.20   
3243            0.0             0.0               0.0                0.00   
2971            0.0             0.0               0.0                0.00   

      word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
3550              0.0            0.21  ...                   0.0         0.00   
1799              0.0            0.00  ...                   0.0         0.00   
3041              0.0            0.20  ...                   0.0         0.00   
3243              0.0            0.00  ...                   0.0         0.00   
2971              0.0            0.00  ...                   0.0         0.08   

      char_freq_(  char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
3550        0.149          0.0        0.000        0.000          0.0   
1799        0.207          0.0        0.207        0.207          0.0   
3041        0.000          0.0        0.060        0.000          0.0   
3243        0.000          0.0        0.000        0.000          0.0   
2971        0.322          0.0        0.000        0.000          0.0   

      capital_run_length_average  capital_run_length_longest  \
3550                   -0.122406                   -0.135798   
1799                   -0.051742                   -0.104616   
3041                   -0.095159                   -0.048488   
3243                   -0.125131                   -0.173217   
2971                   -0.046758                   -0.098380   

      capital_run_length_total  
3550                 -0.144142  
1799                 -0.311754  
3041                 -0.084170  
3243                 -0.414782  
2971                 -0.193349  

[5 rows x 57 columns]

## Create and Fit a Logistic Regression Model

Create a Logistic Regression model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like.

In [ ]:
# Train a Logistic Regression model and print the model score
from sklearn.linear_model import LogisticRegression

logistic_regression_model = LogisticRegression(random_state=1)
logistic_regression_model.fit(X_train_scaled, y_train)

print(f"Training Data Score for Logistical Regression: {logistic_regression_model.score(X_train_scaled, y_train)}")
print(f"Testing Data Score Logistical Regression: {logistic_regression_model.score(X_test_scaled, y_test)}")



Training Data Score for Logistical Regression: 0.9304347826086956
Testing Data Score Logistical Regression: 0.9278887923544744


In [ ]:
# Make and save testing predictions with the saved logistic regression model using the test data
predictions = logistic_regression_model.predict(X_test_scaled)
# Review the predictions

df_results = pd.DataFrame({"Prediction": predictions, "Actual": y_test})
df_results.head()

Prediction  Actual
3550           0       0
1799           1       1
3041           1       0
3243           0       0
2971           0       0

In [ ]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
print(f"The accuracy score of y_test predictions: {accuracy_score(y_test, predictions)}")

The accuracy score of y_test predictions: 0.9278887923544744


## Create and Fit a Random Forest Classifier Model

Create a Random Forest Classifier model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like.

In [ ]:
# Train a Random Forest Classifier model and print the model score
from sklearn.ensemble import RandomForestClassifier
rf_model = RandomForestClassifier(n_estimators=128, random_state=78)
rf_model_fit = rf_model.fit(X_train_scaled, y_train)

print(f"Training Data Score for Random Forest: {rf_model.score(X_train_scaled, y_train)}")
print(f"Testing Data Score Random Forest: {rf_model.score(X_test_scaled, y_test)}")


Training Data Score for Random Forest: 0.9994202898550725
Testing Data Score Random Forest: 0.9496090356211989


In [ ]:
# Make and save testing predictions with the saved logistic regression model using the test data
testing_predictions = rf_model.predict(X_test_scaled)
testing_predictions[:5]
df_testing_predictions = pd.DataFrame({"Prediction": testing_predictions, "Actual": y_test})
df_testing_predictions.head()
# Review the predictions


Prediction  Actual
3550           0       0
1799           1       1
3041           0       0
3243           0       0
2971           0       0

In [ ]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
print(f"The accuracy score of y_test predictions of Random Forest: {accuracy_score(y_test, testing_predictions)}")

The accuracy score of y_test predictions of Random Forest: 0.9496090356211989


## Evaluate the Models

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the following markdown cell.

With an accuaracy score of 0.9496..., the Random Forest Model has a highers score than the Logitical Regression model (scored at 0.9270...).  This is as expected. By layering multiple models, Random Forest is a able to return predictions with higher accuracy.